WIP - working user blueprints in TLD json format

dommrogers opened the pull request:
see #warning TODO items for notes/thoughts

dommrogers committed:
add CI

dommrogers committed:
refactor & adjustments to compile

dommrogers committed:
tabs not spaces

dommrogers committed:
spaces to tab (whole project)

dommrogers committed:
restore OnApplicationLateStart

dommrogers committed:
global::Il2Cpp -> Il2Cpp

dommrogers committed:
#warning bpItem.m_RequiredGearUnits

dommrogers committed:
add Nuget information to project file

dommrogers committed:
add #warning for icon loading

dommrogers committed:
attempt at nuget pkg upload

dommrogers committed:
csproj - update nuget PackageTags

dommrogers committed:
GetRequiredGearItems - remove list and use array directly

dommrogers committed:
GetRequiredGearItems  - fix array content

dommrogers committed:
add #warning for BlueprintData.m_CraftingAudio

dommrogers committed:
use BlueprintManager.LoadUserBlueprint &load from the mod raw JSON

working but there are some TODO items

dommrogers committed:
Merge branch 'round2'

dommrogers committed:
remove PostBuildEvent, breaks CI build

ds5678 commented on CraftingRevisions/BlueprintManager.cs:
```diff
@@ -35,7 +37,12 @@ public static void AddBlueprintFromJson(string text, bool validateEarly)
+				jsonUserBlueprints.Add(text);
+			}
 		}
+
```
Not necessary

ds5678 commented on CraftingRevisions/BlueprintManager.cs:
```diff
@@ -35,7 +37,12 @@ public static void AddBlueprintFromJson(string text, bool validateEarly)
 			}
 
-			AddBlueprint(ModBlueprintData.ParseFromJson(text), validateEarly);
+			// add the blueprint to the HasSet
+			if (!jsonUserBlueprints.Contains(text))
+			{
+				jsonUserBlueprints.Add(text);
+			}
```
You don't need to check if it's contained. Hash sets do that automatically in the add method.

ds5678 commented on CraftingRevisions/BlueprintManager.cs:
```diff
@@ -5,6 +5,8 @@ public static class BlueprintManager
 		private static readonly List<ModBlueprintData> pendingBlueprints = new List<ModBlueprintData>();
 		private static bool registeredPendingBlueprints;
 
+		public static HashSet<string> jsonUserBlueprints = new();
```
private or internal, but not public

ds5678 commented on CraftingRevisions/CraftingRevisions.csproj:
```diff
@@ -1,4 +1,4 @@
-﻿<Project Sdk="Microsoft.NET.Sdk">
+<Project Sdk="Microsoft.NET.Sdk">
```
What's going on here?

dommrogers replied:
not sure, can't locate this is any of my commits

ds5678 commented on CraftingRevisions/Patches/GameManager_Awake.cs:
```diff
@@ -4,15 +4,41 @@
+					// load the blueprint into the game
+					bool loaded = __instance.LoadUserBlueprint(jsonUserBlueprint);
+
+#warning TODO - there appears to be no way to define an icon for this blueprint (UserBlueprintData still a WIP?)
```
Icons are loaded with a naming convention based on the gear result name. It was the same before the DLC.

dommrogers replied:
icons are working, but appears to only show for certain gearitems (e.g. bandage is fine but Prybar is not)

ds5678 commented on CraftingRevisions/Patches/GameManager_Awake.cs:
```diff
@@ -4,15 +4,41 @@
+#warning TODO - there appears to be no way to define an icon for this blueprint (UserBlueprintData still a WIP?)
+#warning TODO - get CraftingAudio working, is a string but unclear what value it wants. (see UserBlueprintData.MakeRuntimeWwiseEvent)
+#warning TODO - work on loading validation if it fails
+#warning TODO - Should we attempt to detect & convert "old" format json ?
```
Not in the mod.

ds5678 commented on CraftingRevisions/Patches/GameManager_Awake.cs:
```diff
@@ -4,15 +4,41 @@
+					bool loaded = __instance.LoadUserBlueprint(jsonUserBlueprint);
+
+#warning TODO - there appears to be no way to define an icon for this blueprint (UserBlueprintData still a WIP?)
+#warning TODO - get CraftingAudio working, is a string but unclear what value it wants. (see UserBlueprintData.MakeRuntimeWwiseEvent)
```
I don't understand why this is an issue. Are the old strings not working?

dommrogers replied:
I've tried to debug this quite a bit, when the UserBlueprintData is loaded it calls MakeRuntimeWwiseEvent with the string but it throws an exception.

```
 During invoking native->managed trampoline
[22:43:48.206] Il2CppInterop.Runtime.Il2CppException: System.NullReferenceException: Object reference not set to an instance of an object.
--- BEGIN IL2CPP STACK TRACE ---
System.NullReferenceException: Object reference not set to an instance of an object.
  at TLD.Gear.UserBlueprintData.MakeRuntimeWwiseEvent (System.String eventName) [0x00000] in <00000000000000000000000000000000>:0
--- END IL2CPP STACK TRACE ---

   at Il2CppInterop.Runtime.Il2CppException.RaiseExceptionIfNecessary(IntPtr returnedException) in /home/runner/work/Il2CppInterop/Il2CppInterop/Il2CppInterop.Runtime/Il2CppException.cs:line 36
   at DMD<Il2CppTLD.Gear.UserBlueprintData::MakeRuntimeWwiseEvent>(String eventName)
   at (il2cpp -> managed) MakeRuntimeWwiseEvent(IntPtr , Il2CppMethodInfo* )
```

```
[Cpp2ILInjected.CallAnalysis.Calls(Type = typeof(AkSoundEngine), Member = "GetIDFromString")]
```
** Somewhere after here as I can patch the postfix of GetIDFromString and get the result >>
[22:43:48.203] AkSoundEngine_GetIDFromString_POST | PLAY_CRAFTINGCLOTH
[22:43:48.203] AkSoundEngine_GetIDFromString_POST | 1661985380
```
[Cpp2ILInjected.CallAnalysis.Calls(Type = typeof(ScriptableObject), Member = "CreateInstance")]
[Cpp2ILInjected.CallAnalysis.Calls(Type = typeof(Type), Member = "GetTypeFromHandle")]
[Cpp2ILInjected.CallAnalysis.Calls(Type = typeof(FieldInfo), Member = "SetValue")]
[Cpp2ILInjected.CallAnalysis.CallsDeduplicatedMethods(Count = 3)]
```

I'll continue to dig...

ds5678 replied:
> calls MakeRuntimeWwiseEvent with the string but it throws an exception.

I think this is caused by:
```json
"CraftingAudio":null,
```
Where the proper value would be:
```json
"CraftingAudio":"",
```
Because `MakeRuntimeWwiseEvent` might not support null input values.

> I'll continue to dig...

https://melonwiki.xyz/#/modders/debugging

Disclaimer: I've never done this for modding, but it's very useful for normal software development.

ds5678 commented on CraftingRevisions/TestBlueprints.cs:
```diff
@@ -0,0 +1,59 @@
+
+		public static string blueprint_2 = """
+			{
+			"Name":"UserBp_Prybar",
```
Is name part of the Hinterland blueprints, or something you added for use in validation validation, similar to the old format?

ds5678 commented on CraftingRevisions/TestBlueprints.cs:
```diff
@@ -0,0 +1,59 @@
+
+namespace CraftingRevisions
+{
+	internal class TestBlueprints
```
FYI, I'm likely going to want this class removed before merging can happen.

ds5678 commented on CraftingRevisions/TestBlueprints.cs:
```diff
@@ -0,0 +1,59 @@
+	internal class TestBlueprints
+	{
+
+		public static string blueprint_1 = """
```
Just a general tip: `public const string` is better

ds5678 replied:
Raw string literals are nice

dommrogers committed:
internal HashSet & remove .Contains check

dommrogers committed:
Remove "Name" from test blueprint_2

Was thinking it could be used as part of validation output to identify which blueprint fails.

dommrogers committed:
add #warning comment to TestBlueprints

To not forget it needs to be removed for merge&release

dommrogers committed:
add few #warning  regarding removing logic before merge&release

+minor format cleanup of test blueprints

dommrogers committed:
Strip old unused code

dommrogers committed:
remove additional unused file

dommrogers committed:
override UserBlueprintData.MakeRuntimeWwiseEvent

Appears broken/incomplete, the bypass works without error and audio plays fine.

dommrogers committed:
move #warning about LoadAllUserBlueprints

ds5678 commented on CraftingRevisions/Patches/GameManager_Awake.cs:
```diff
@@ -11,11 +11,11 @@ internal class Panel_Crafting_Initialize
 		private static void Postfix()
 		{
 			// call TLD LoadAllUserBlueprints
+#warning BlueprintManager.LoadAllUserBlueprints is not called by the game itself yet, we may need to remove this to prevent dupe recipes in the future.
```
It won't cause duplicate recipes. User recipes are cleared whenever this is called.

ds5678 commented on CraftingRevisions/Patches/GameManager_Awake.cs:
```diff
@@ -32,13 +32,29 @@ private static void Postfix(Il2CppTLD.Gear.BlueprintManager __instance)
+	[HarmonyPatch(typeof(Il2CppTLD.Gear.UserBlueprintData), nameof(Il2CppTLD.Gear.UserBlueprintData.MakeRuntimeWwiseEvent), new Type[] { typeof(string) })]
+	internal class UserBlueprintData_MakeRuntimeWwiseEvent
+	{
+		private static void Prefix(string eventName, ref bool __runOriginal)
```
eventName is unnecessary and creates lots of string marshalling.

ds5678 commented on CraftingRevisions/BlueprintManager.cs:
```diff
@@ -14,5 +16,24 @@ public static void AddBlueprintFromJson(string text, bool validateEarly)
 			jsonUserBlueprints.Add(text);
 		}
+
+		public static Il2CppAK.Wwise.Event? MakeAudioEvent(string eventName)
```
Golden rule of API design: if it doesn't need to be public, don't make it public.

eventName should be nullable since you allow null values.

ds5678 commented on CraftingRevisions/Patches/GameManager_Awake.cs:
```diff
@@ -32,13 +32,29 @@ private static void Postfix(Il2CppTLD.Gear.BlueprintManager __instance)
+		}
+		private static void Postfix(string eventName, ref bool __runOriginal, ref Il2CppAK.Wwise.Event? __result)
+		{
+			__result = BlueprintManager.MakeAudioEvent(eventName);
```
Move MakeAudioEvent to this class. I don't see any reason for it to be exposed to the rest of the mod.

ds5678 commented:
Also, I think you have the power to resolve conversations. You should do that whenever you change something.

dommrogers committed:
WIP validation

dommrogers committed:
Il2CppAK.Wwise.Event emptyEvent

dommrogers committed:
PR #3 changes

dommrogers committed:
cleanup & improved validation output

ds5678 commented on CraftingRevisions/BlueprintManager.cs:
```diff
@@ -1,41 +1,28 @@
-﻿namespace CraftingRevisions
+﻿
+namespace CraftingRevisions
```
Remove the empty line.

ds5678 commented on CraftingRevisions/UserBlueprintData.cs:
```diff
@@ -0,0 +1,178 @@
+
+		public List<RequiredGearItem> RequiredGear;
+
+		// Token: 0x04007043 RID: 28739
```
Remove all of these.

FYI, you can disable these in your decompiler settings.

ds5678 commented on CraftingRevisions/UserBlueprintData.cs:
```diff
@@ -0,0 +1,178 @@
+
+		internal void Validate()
+		{
+			string exception = "";
```
Use System.AggregateException instead.

ds5678 commented on CraftingRevisions/UserBlueprintData.cs:
```diff
@@ -0,0 +1,178 @@
+		public int Count;
+	}
+
+	public enum ModSkillType
+	{
+		None = -1,
+		Firestarting,
+		CarcassHarvesting,
+		IceFishing,
+		Cooking,
+		Rifle,
+		Archery,
+		ClothingRepair,
+		ToolRepair,
+		Revolver,
+		Gunsmithing
+	}
+
+	public enum ModCraftingLocation
+	{
+		Anywhere,
+		Workbench,
+		Forge,
+		AmmoWorkbench
+	}
```
Use the Hinterland enums if they're the same.

ds5678 commented on CraftingRevisions/UserBlueprintData.cs:
```diff
@@ -0,0 +1,178 @@
+		AmmoWorkbench
+	}
+}
+
+
```
Remove

ds5678 commented on CraftingRevisions/UserBlueprintData.cs:
```diff
@@ -0,0 +1,178 @@
+
+namespace CraftingRevisions
+{
+	public class ModUserBlueprintData
```
internal sealed class

ds5678 commented on CraftingRevisions/UserBlueprintData.cs:
```diff
@@ -0,0 +1,178 @@
+		/// </summary>
+		public string? Name = null;
+
+		public List<RequiredGearItem> RequiredGear;
```
Nullable

ds5678 commented on CraftingRevisions/UserBlueprintData.cs:
```diff
@@ -0,0 +1,178 @@
+		/// <summary>
+		/// optional name, used for debugging
+		/// </summary>
+		public string? Name = null;
```
Only ever initialize a field if you're not initializing to the default value for that data type (null or zero).

ds5678 commented on CraftingRevisions/UserBlueprintData.cs:
```diff
@@ -0,0 +1,178 @@
+		public List<string>? OptionalTools = new();
+
+		// Token: 0x04007045 RID: 28741
+		public string CraftedResult;
```
Nullable

ds5678 commented on CraftingRevisions/UserBlueprintData.cs:
```diff
@@ -0,0 +1,178 @@
+		public List<RequiredGearItem> RequiredGear;
+
+		// Token: 0x04007043 RID: 28739
+		public string? RequiredTool = null;
```
All the fields in this class should be made into auto properties.

ds5678 commented on CraftingRevisions/UserBlueprintData.cs:
```diff
@@ -0,0 +1,178 @@
+		}
+		#endregion
+
+		public string GetName() => Name ?? "";
```
This seems unnecessary. String interpolation doesn't choke on null values.

ds5678 commented on CraftingRevisions/UserBlueprintData.cs:
```diff
@@ -0,0 +1,178 @@
+		#region Json
+		public static ModUserBlueprintData ParseFromJson(string jsonText)
+		{
+			return JSON.Load(jsonText).Make<ModUserBlueprintData>();
```
System.Text.Json

Here's some code for reference.
https://github.com/AssetRipper/AssetRipper/blob/master/Source/AssetRipper.Export.UnityProjects/Scripts/AssemblyDefinitions/AssemblyDefinitionSerializerContext.cs
https://github.com/AssetRipper/AssetRipper/blob/master/Source/AssetRipper.Export.UnityProjects/Scripts/AssemblyDefinitions/AssemblyDefinitionExporter.cs#L27

ds5678 commented on CraftingRevisions/UserBlueprintData.cs:
```diff
@@ -0,0 +1,178 @@
+		// Token: 0x04007054 RID: 28756
+		public bool CanIncreaseRepairSkill = false;
+
+		#region Json
```
A region isn't necessary when it only contains one method.

ds5678 commented on CraftingRevisions/UserBlueprintData.cs:
```diff
@@ -0,0 +1,178 @@
+		/// <summary>
+		/// String value of the gear item
+		/// </summary>
+		public string Item;
```
Nullable

ds5678 commented on CraftingRevisions/Patches/GameManager_Awake.cs:
```diff
@@ -1,19 +1,99 @@
+			__runOriginal = false;
+		}
+
+		internal static Il2CppAK.Wwise.Event? MakeAudioEvent(string? eventName)
```
private static

Also, I like that you moved this here. Encapsulation is good.

ds5678 commented on CraftingRevisions/Patches/GameManager_Awake.cs:
```diff
@@ -1,19 +1,99 @@
+			newEvent.WwiseObjectReference.id = eventId;
+			return newEvent;
+		}
+		internal static uint GetAKEventIdFromString(string eventName)
```
private static

ds5678 commented on CraftingRevisions/Patches/GameManager_Awake.cs:
```diff
@@ -1,19 +1,99 @@
+		private static void Postfix(Il2CppTLD.Gear.BlueprintManager __instance)
+		{
+			// check we have a HasSet and it has contents
+			if (BlueprintManager.jsonUserBlueprints != null && BlueprintManager.jsonUserBlueprints.Count > 0)
```
This if statement isn't needed. The hash set will never be null, and checking the size is redundant.

ds5678 commented on CraftingRevisions/Patches/GameManager_Awake.cs:
```diff
@@ -1,19 +1,99 @@
+					// load the blueprint into the game
+					bool loaded = __instance.LoadUserBlueprint(jsonUserBlueprint);
+
+					if (loaded == false)
```
Don't use equality comparisons for boolean values when one side is constant. Use the ! operator instead.

dommrogers committed:
PR Changes

dommrogers committed:
Non-nullable field x 2 fixed

ds5678 commented on CraftingRevisions/ModUserBlueprintData.cs:
```diff
@@ -0,0 +1,126 @@
+				throw new AggregateException("Aggregate Exception Message", exceptions);
+
+			}
+			catch (AggregateException ae)
+			{
+				MelonLoader.MelonLogger.Error(ae.Message);
+			}
```
This is a distinct change in behavior from what you had before. Before you were throwing the exception without catching it, and I suggested the AggregateException because of that.

Using exceptions for control flow is extremely bad and should almost never be done. Use a StringBuilder to make your error message.

dommrogers replied:
I went with the information I had `Use System.AggregateException instead.` and this is how I found most recommend using it (please don't forget C# is not my primary language)

I'f you have any other suggestions on how to use it better I'm open to them.

ds5678 replied:
Let's not use exceptions here. Error logging is sufficient.

Use `System.Text.StringBuilder` to construct your error message.

ds5678 commented on CraftingRevisions/ModUserBlueprintData.cs:
```diff
@@ -0,0 +1,126 @@
+		public SkillType ImprovedSkill = SkillType.None;
+		public CraftingLocation RequiredCraftingLocation = CraftingLocation.Anywhere;
+		public bool RequiresLitFire = false;
+		public bool CanIncreaseRepairSkill = false;
```
Convert all these to auto properties.

ds5678 commented on CraftingRevisions/ModUserBlueprintData.cs:
```diff
@@ -0,0 +1,126 @@
+		#region Json
+		public static ModUserBlueprintData ParseFromJson(string jsonText)
+		{
+			return JSON.Load(jsonText).Make<ModUserBlueprintData>();
```
Use System.Text.Json

https://github.com/AssetRipper/AssetRipper/blob/master/Source/AssetRipper.Export.UnityProjects/Scripts/AssemblyDefinitions/AssemblyDefinitionExporter.cs#L27
https://github.com/AssetRipper/AssetRipper/blob/master/Source/AssetRipper.Export.UnityProjects/Scripts/AssemblyDefinitions/AssemblyDefinitionSerializerContext.cs

ds5678 commented on CraftingRevisions/ModUserBlueprintData.cs:
```diff
@@ -0,0 +1,126 @@
+		}
+		#endregion
+
+		public string GetName() => Name ?? "";
```
Remove. `Name` can be used directly in the string interpolations below. `?? ""` isn't necessary for the string interpolations because they already do their own null checking.

ds5678 commented on CraftingRevisions/ModUserBlueprintData.cs:
```diff
@@ -0,0 +1,126 @@
+		}
+		#endregion
+
+		private static T[] GetEnumValues<T>() where T : Enum
+		{
+			return (T[])(Enum.GetValues(typeof(T)));
+		}
+
+		private static bool IsValidEnumValue<T>(T value) where T : Enum
+		{
+			return GetEnumValues<T>().Contains(value);
+		}
```
```cs
private static class EnumValues<T> where T : Enum
{
    private static readonly T[] values = Enum.GetValues<T>();
    public static bool Contains(T value) => values.Contains(value);
}
```

dommrogers replied:
A bit more context please, the above is not valid.
`The type 'T' must be a non-nullable value type in order to use it as parameter 'TEnum' in the generic type or method 'Enum.GetValues<TEnum>()'`

ds5678 replied:
```cs
private static class EnumValues<T> where T : struct, Enum
{
    private static readonly T[] values = Enum.GetValues<T>();
    public static bool Contains(T value) => values.Contains(value);
}
```

ds5678 commented on CraftingRevisions/ModUserBlueprintData.cs:
```diff
@@ -0,0 +1,126 @@
+
+namespace CraftingRevisions
+{
+	public class ModUserBlueprintData
```
internal sealed class

ds5678 replied:
Bump

ds5678 commented on CraftingRevisions/ModUserBlueprintData.cs:
```diff
@@ -0,0 +1,126 @@
+
+	}
+
+	public class RequiredGearItem
```
internal sealed class

ds5678 commented on CraftingRevisions/Patches/GameManager_Awake.cs:
```diff
@@ -1,19 +1,90 @@
+	[HarmonyPatch(typeof(Il2CppTLD.Gear.UserBlueprintData), nameof(Il2CppTLD.Gear.UserBlueprintData.MakeRuntimeWwiseEvent), new Type[] { typeof(string) })]
+	internal class UserBlueprintData_MakeRuntimeWwiseEvent
+	{
+		private static void Prefix(ref bool __runOriginal)
+		{
+			__runOriginal = false;
+		}
+		private static void Postfix(string eventName, ref bool __runOriginal, ref Il2CppAK.Wwise.Event? __result)
+		{
+			__result = MakeAudioEvent(eventName);
+			__runOriginal = false;
+		}
```
Two scenarios:

If the game throws an exception whenever the original is run:

* Move the code into the prefix and delete the postfix.

If the game gracefully returns null whenever the original is run:

* Delete the prefix and only make the event whenever Hinterland doesn't.
* If it *always* returns null, add a log statement to remind us in the future when this gets implemented.

ds5678 commented on CraftingRevisions/Patches/GameManager_Awake.cs:
```diff
@@ -1,19 +1,90 @@
+		}
+		private static Il2CppAK.Wwise.Event? MakeAudioEvent(string? eventName)
+		{
+			Il2CppAK.Wwise.Event emptyEvent = new();
+			emptyEvent.WwiseObjectReference = ScriptableObject.CreateInstance<WwiseEventReference>();
+
+			if (eventName == null || eventName == "")
+			{
+				return emptyEvent;
+			}
+			uint eventId = GetAKEventIdFromString(eventName);
+			if (eventId == 0U)
+			{
+				return emptyEvent;
+			}
```
```cs
if (string.IsNullOrEmpty(eventName) || GetAKEventIdFromString(eventName) == 0)
{
    Il2CppAK.Wwise.Event emptyEvent = new();
    emptyEvent.WwiseObjectReference = ScriptableObject.CreateInstance<WwiseEventReference>();
    return emptyEvent;
}
```

ds5678 commented on CraftingRevisions/Patches/GameManager_Awake.cs:
```diff
@@ -1,19 +1,90 @@
+				return emptyEvent;
+			}
+
+			Il2CppAK.Wwise.Event newEvent = new();
+			newEvent.WwiseObjectReference = ScriptableObject.CreateInstance<WwiseEventReference>();
+			newEvent.WwiseObjectReference.objectName = eventName;
+			newEvent.WwiseObjectReference.id = eventId;
+			return newEvent;
```
```cs
return new()
{
    WwiseObjectReference = ScriptableObject.CreateInstance<WwiseEventReference>(),
    WwiseObjectReference.objectName = eventName, //I'm not sure if this line is valid.
    WwiseObjectReference.id = eventId,
};
```

dommrogers replied:
Not sure how to go about getting this to work `Invalid initializer member declarator` for both objectName and id 

Also we want the `Il2CppAK.Wwise.Event` as a return not the `WwiseObjectReference`

ds5678 replied:
Then don't worry about it.

ds5678 commented on CraftingRevisions/Patches/GameManager_Awake.cs:
```diff
@@ -1,19 +1,90 @@
+		private static uint GetAKEventIdFromString(string eventName)
+		{
+			Type type = typeof(Il2CppAK.EVENTS);
+			foreach (PropertyInfo prop in type.GetProperties(BindingFlags.Static | BindingFlags.Public | BindingFlags.Instance))
 			{
-				MelonLogger.Error($"Blueprint Mapping failed: {e}");
+				if (prop.Name.ToLower() == eventName.ToLower())
+				{
+					return Convert.ToUInt32(prop.GetValue(null)?.ToString());
+				}
 			}
+			return 0U;
```
Two things:

- Build a dictionary and cache it.
- ToLowerInvariant.

ds5678 commented on CraftingRevisions/ModUserBlueprintData.cs:
```diff
@@ -0,0 +1,126 @@
+		/// <summary>
+		/// String value of the gear item
+		/// </summary>
+		public string? Item = null;
+		/// <summary>
+		/// Count of how many are required
+		/// </summary>
+		public int Count = 0;
```
Make these into auto properties.

ds5678 commented on CraftingRevisions/Patches/GameManager_Awake.cs:
```diff
@@ -1,19 +1,90 @@
+		}
+	}
+
+	// patched into postfix BlueprintManager.LoadAllUserBlueprints
+	// (BlueprintManager.RemoveUserBlueprints is not guaranteed to get called)
+	[HarmonyPatch(typeof(Il2CppTLD.Gear.BlueprintManager), nameof(Il2CppTLD.Gear.BlueprintManager.LoadAllUserBlueprints))]
+	internal class BlueprintManager_LoadAllUserBlueprints_Postfix
+	{
+		private static void Postfix(Il2CppTLD.Gear.BlueprintManager __instance)
+		{
+			// loop over the items
+			foreach (string jsonUserBlueprint in BlueprintManager.jsonUserBlueprints)
+			{
+
+				// load the blueprint into the game
+				bool loaded = __instance.LoadUserBlueprint(jsonUserBlueprint);
+
+				if (!loaded)
+				{
+					// validate the blueprint
+					BlueprintManager.ValidateJsonBlueprint(jsonUserBlueprint);
+				}
+			}
+		}
+	}
```
Move this patch to BlueprintManager and use the more concise format. Change jsonUserBlueprints back to private.

ds5678 reviewed (changes requested):
Looking good so far. Whenever you're ready for me to review it again, just let me know.

dommrogers committed:
PR Changes (See PR comments)

ds5678 commented on CraftingRevisions/ModUserBlueprintData.cs:
```diff
@@ -0,0 +1,126 @@
+
+	}
+
+	public class ModRequiredGearItem
```
`internal sealed class ModRequiredGearItem`

ds5678 commented on CraftingRevisions/Patches/GameManager_Awake.cs:
```diff
@@ -1,19 +1,71 @@
+				foreach (PropertyInfo prop in type.GetProperties(BindingFlags.Static | BindingFlags.Public))
+				{
+					string key = prop.Name.ToLowerInvariant();
+					uint value = Convert.ToUInt32(prop.GetValue(null).ToString());
```
```cs
uint value = (uint)prop.GetValue(null)!;
```

ds5678 commented on CraftingRevisions/Patches/GameManager_Awake.cs:
```diff
@@ -1,19 +1,71 @@
+				}
 			}
+
+			string eventKey = eventName.ToLowerInvariant();
+
+			if (eventIds.ContainsKey(eventKey))
+			{
+				return eventIds[eventKey];
+			}
+
+			return 0U;
```
```cs
eventIds.TryGetValue(eventName.ToLowerInvariant(), out uint id);
return id;
```

dommrogers committed:
PR Changes

ds5678 commented on CraftingRevisions/ModUserBlueprintData.cs:
```diff
@@ -0,0 +1,122 @@
+		#region Json
+		public static ModUserBlueprintData ParseFromJson(string jsonText)
+		{
+			return JsonSerializer.Deserialize<ModUserBlueprintData>(jsonText);
```
`?? throw new ArgumentException("Could not parse blueprint data from the text.", nameof(jsonText))`

ds5678 commented on CraftingRevisions/ModUserBlueprintData.cs:
```diff
@@ -0,0 +1,122 @@
+
+	}
+
+	internal sealed class ModRequiredGearItem
+	{
+		/// <summary>
+		/// String value of the gear item
+		/// </summary>
+		public string? Item { get; set; } = null;
+		/// <summary>
+		/// Count of how many are required
+		/// </summary>
+		public int Count { get; set; } = 0;
+	}
```
Move this to its own file.

ds5678 commented on CraftingRevisions/ModUserBlueprintData.cs:
```diff
@@ -0,0 +1,122 @@
+		public int Count { get; set; } = 0;
+	}
+}
+
+
+
+
```
Remove trailing new lines

ds5678 reviewed (changes requested):
Looking great. It should be good to merge after these changes are implemented.

dommrogers committed:
more changes
